### PR TITLE
[2.9.0] 1370 optional filter on azuread auth group memberships

### LINKED ISSUE #1370: [v2.9.0] optional filter on AzureAD auth group memberships
## Related Issues

https://github.com/rancher/rancher/issues/42940
https://jira.suse.com/browse/SURE-5641

## Summary
Created filter so that only groups that begins with a specific name are being returned instead of all the groups the user has attached to his account.

thus, reducing the noise in the logs

"we have many users that have a lot of groups and we have discovered that it generates a lot of logs because all of their groups are displayed in the logs"

## Details
The backend, specifically the AzureADConfig structure is extended with a new field groupMembershipFilter.
The field takes a string and passes this directly into filter clause of the group membership query.
This means that the field has to contain a filter clause in OData syntax.
Example: startswith(userPrincipalName,'fresh')
See also https://learn.microsoft.com/en-us/odata/concepts/queryoptions-overview#filter

### PR DESCRIPTION
**Pull request by martyav**
Fixes #1370 

## Reminders

- See the [README](../README.md) for more details on how to work with the Rancher docs.

- Verify if changes pertain to other versions of Rancher. If they do, finalize the edits on one version of the page, then apply the edits to the other versions.

- If the pull request is dependent on an upcoming release, remember to add a "MERGE ON RELEASE" label and set the proper milestone.

## Description

From the issue:

> Created filter so that only groups that begins with a specific name are being returned instead of all the groups the user has attached to his account.
>
> thus, reducing the noise in the logs
>
> "we have many users that have a lot of groups and we have discovered that it generates a lot of logs because all of their groups are displayed in the logs"
>
> **Details**
>
> The backend, specifically the AzureADConfig structure is extended with a new field groupMembershipFilter.
> The field takes a string and passes this directly into filter clause of the group membership query.
> This means that the field has to contain a filter clause in OData syntax.
> Example: startswith(userPrincipalName,'fresh')
> See also https://learn.microsoft.com/en-us/odata/concepts/queryoptions-overview#filter

## Comments

I included similar content in two places to cover the case of applying the filter during setup, and after setup.

The second case received a longer warning box with more explanation, as it exists in its own section and doesn't interrupt a larger set of instructions.

### CODE REVIEW

File: docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md [hunk 1]
```diff
@@ -221,6 +221,18 @@ To complete configuration, enter information about your AD instance in the Ranch
...
 
+1. (Optional) In Rancher v2.9.0 and later, you can filter user group memberships in Azure AD to reduce the amount of log data generated. 
+    
+    :::warning
+    
+    Filtering out a user group membership can have the side effect of denying users permissions they should have.
+
+    :::
+    
+    1. Click the checkbox next to **Limit users by group membership**.
```
Comment (anupama2501) [inline docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md:232]: This seems repetitive provided that its already defined well in step 329. Can we remove this part and only have the part added in lines 329 through 349?

File: docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md [hunk 2]
```diff
@@ -314,6 +326,27 @@ Endpoint         | https://login.partner.microsoftonline.cn/
...
+
+:::
+
+1. In Rancher, in the top left corner, click **☰ > Users & Authentication**.
+
+1. In the left navigation menu, click **Auth Provider**.
+
+1. Click **AzureAD**.
+
+1. Click the checkbox next to **Limit users by group membership**.
```
Comment (anupama2501) [inline docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md:347]: can we add an image showing the filter here?

File: docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md [hunk 3]
```diff
@@ -314,6 +326,27 @@ Endpoint         | https://login.partner.microsoftonline.cn/
...
+    
+1. Enter an [OData filter clause](https://learn.microsoft.com/en-us/odata/concepts/queryoptions-overview#filter) into the **Group Membership Filter** field. For example, if you want to limit logging to group memberships whose name starts with `Rancher`, click the checkbox and enter `startswith(displayName,'Rancher')`.
```
[the remark this answers is not in the document]
Comment (martyav) [reply docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md:349]: The Microsoft link provides a few examples, and links to the main spec (which might be overkill, but we could also link here -- it's at https://docs.oasis-open.org/odata/odata/v4.01/cs01/part2-url-conventions/odata-v4.01-cs01-part2-url-conventions.html#sec_SystemQueryOptionfilter)

[the remark this answers is not in the document]
Comment (btat) [reply docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md:339]: The links you're listing are only for the filters (unless it's buried deep and I'm glossing over them). I'm referring to the parameters you'd pass to those filters. Using the `startswith(displayName,'Rancher')` example on the page, how does a user know `displayName`. written in camel case, is a valid parameter to pass to the filter. If they want to filter using a field other than `displayName`, where would they find those?

If we're assuming this is common knowledge for someone using Azure AD we can ignore.

[the remark this answers is not in the document]
Comment (martyav) [reply docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-azure-ad.md]: It seems like Azure calls these fields "attributes". I'll see if they have any straightforward documentation on working with them.

**Review by btat — APPROVED** [review]

**Pull request merged**